src/core/lib/slice/slice_internal.h

```diff
@@ -215,22 +215,28 @@ struct InternedSliceRefcount {
 }  // namespace grpc_core
 
 inline int grpc_slice_refcount::Eq(const grpc_slice& a, const grpc_slice& b) {
+  GPR_DEBUG_ASSERT(a.refcount != nullptr);
+  GPR_DEBUG_ASSERT(a.refcount == this);
   switch (ref_type_) {
     case Type::STATIC:
-      return GRPC_STATIC_METADATA_INDEX(a) == GRPC_STATIC_METADATA_INDEX(b);
+      GPR_DEBUG_ASSERT(
+          (GRPC_STATIC_METADATA_INDEX(a) == GRPC_STATIC_METADATA_INDEX(b)) ==
+          (a.refcount == b.refcount));
     case Type::INTERNED:
       return a.refcount == b.refcount;
     case Type::NOP:
     case Type::REGULAR:
       break;
   }
-  if (GRPC_SLICE_LENGTH(a) != GRPC_SLICE_LENGTH(b)) return false;
-  if (GRPC_SLICE_LENGTH(a) == 0) return true;
-  return 0 == memcmp(GRPC_SLICE_START_PTR(a), GRPC_SLICE_START_PTR(b),
-                     GRPC_SLICE_LENGTH(a));
+  if (a.data.refcounted.length != GRPC_SLICE_LENGTH(b)) return false;
+  if (a.data.refcounted.length == 0) return true;
+  return 0 == memcmp(a.data.refcounted.bytes, GRPC_SLICE_START_PTR(b),
```

Comment:
why do we have to replace the macros here?

Reply:
The macros are basically a branch on whether the slice is inlined (ie. refcount is null) or not.
However, we don't need to do that branch since if we're here, we *know* we're not inlined.

Reply:
I see but, instead of removing the macros, I would suggest adding inlines function that has `grpc_refcounted_slice_length()` and `grpc_ref_counted_slice_data()` with debug asserts that `refcount` is not nullptr, for readability.

Reply:
Done.